### ISSUE TITLE
realloc(ptr, 0) ought not behave like free()

### ISSUE THREAD

**Issue by emaste** [jemalloc/jemalloc#1629]
prompted by the discussion following https://twitter.com/RichFelker/status/1171108501751554048 we plan to have FreeBSD return non-NULL for realloc(ptr, 0) as malloc(0) does

From the comment in the test case at https://sortix.org/os-test/malloc/realloc-0.c:
```
		/* realloc returns NULL without setting errno. That means the allocation
		   been freed and we didn't get a replacement allocation. This behavior
		   is undesirable in my opinion because it causes much more compexity
		   and makes realloc much harder to use without a check for whether size
		   is zero. Unfortunately C11 with DR400
		   <http://open-std.org/jtc1/sc22/wg14/www/docs/summary.htm#dr_400>
		   now allows this behavior and marks using realloc with size == 0 as
		   obsolescent. POSIX issue 7 (2018) also allows this behavior, even
		   though its rationale doesn't like it. It does require errno to be set
		   to an implementation specific value in this case, which no
		   implementation does, so I guess that means keeping errno unchanged.
		   Therefore this case is allowed by the standards. It does make the
		   interface much harder to use for arbitrary lengths and can cause
		   double free and use after free bugs if software doesn't know to take
		   care. */
```

FreeBSD PR: https://bugs.freebsd.org/240456
FreeBSD change review: https://reviews.freebsd.org/D21632

**Comment by davidtgoldblatt** [comment]
Summarizing an offline discussion for posterity / openness: the safety/compliance benefits are probably worth risking some leaks in screwy realloc-as-free programs. But since it introduces leaks to previously unleaky programs expecting glibc/previous-jemalloc semantics, we don't want to abandon those people. So we can make the behavior a malloc_conf option; something like `zero_realloc:compliant|free|abort`. Programs that want the current semantics can set `zero_realloc:free`, and those that want to check that they're portable across all libcs (and thus never call realloc with a size of zero) can specify `zero_realloc:abort`. FreeBSD can have it set to compliant by default, but stuff in ports can still change behavior during early startup.

**Comment by davidtgoldblatt** [comment]
A tricky question we didn't talk about was what to do in the default configuration. Some possibilities:
- `zero_realloc:compliant` by default. Using realloc-as-free can be punished with leaks, but not with security vulnerabilities.
- Try to do what the platform allocator does (I don't know what we'd do for cross-compilation though)
- `zero_realloc:free` by default. If someone's workflow step is "download and compile jemalloc without caring about how it's configured", they're probably trying it as a drop-in for glibc.

**Comment by davidtgoldblatt** [comment]
Fixed in #1643.
